batch/sql/estimated-current.sql

```diff
@@ -1205,18 +1215,19 @@ BEGIN
     SET state = new_state, status = new_status, attempt_id = in_attempt_id
     WHERE batch_id = in_batch_id AND job_id = in_job_id;
 
-    UPDATE batches SET n_completed = n_completed + 1 WHERE id = in_batch_id;
-    UPDATE batches
-      SET time_completed = new_timestamp,
-          `state` = 'complete'
-      WHERE id = in_batch_id AND n_completed = batches.n_jobs;
+    UPDATE batches_n_jobs_in_complete_states
+      SET n_completed = n_completed + 1
+      WHERE id = in_batch_id;
+    SELECT n_completed INTO new_n_completed
+    FROM batches_n_jobs_in_complete_states
+    WHERE id = in_batch_id;
```

Comment:
You've dropped the modification of the state of a batch, which I think is a bug.

Reply:
In particular, for this to work I think you *must* move `state` into the volatile table.

I think the only things that a job should be permitted to place shared read locks on are:
- the user which owns the batch
- the billing project in which a batch resides
- whether a batch is cancelled
The first two are immutable properties of a batch.

Reply:
Ah, I see you moved the update to state into a separate transaction. I don't think that will help much.

I'm not sure we can do this, but if we can move the update of the parent jobs to before the update to the volatile state, then we'll only lock the batch row for a very short period of time at the very end of MJC. That shouldn't cause deadlocks and should have a small effect on latency.

In particular I imagine this:

```sql
UPDATE batches_n_jobs_in_complete_states, batches
SET n_completed = n_completed + 1,
     n_cancelled = n_cancelled + (new_state = 'Cancelled'),
     n_failed = n_failed + (new_state = 'Error' OR new_state = 'Failed'),
     n_succeeded = n_succeeded + (new_state != 'Cancelled' AND new_state != 'Error' AND new_state != 'Failed'),
    state = IF(n_completed + 1 = batches.n_jobs, 'complete', state)
WHERE id = in_batch_id
```